oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,463 @@
+    long currentTime = clock.currentTimeMillis();
+    String assertion = createAssertion(jsonFactory, currentTime, getApiAudience());
+
+    GenericData tokenRequest = new GenericData();
+    tokenRequest.set("grant_type", GRANT_TYPE);
+    tokenRequest.set("assertion", assertion);
+    UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
+
+    HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
+    HttpRequest request = requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), content);
+
+    request.setParser(new JsonObjectParser(jsonFactory));
+
+    HttpResponse response;
+    String errorTemplate = "Error getting access token for GDCH service account: %s, iss: %s";
+
+    try {
+      response = request.execute();
+    } catch (HttpResponseException re) {
+      String message = String.format(errorTemplate, re.getMessage(), getServiceIdentityName());
+      throw GoogleAuthException.createWithTokenEndpointResponseException(re, message);
+    } catch (IOException e) {
+      throw GoogleAuthException.createWithTokenEndpointIOException(
+          e, String.format(errorTemplate, e.getMessage(), getServiceIdentityName()));
+    }
+
+    GenericData responseData = response.parseAs(GenericData.class);
+    String accessToken =
+        OAuth2Utils.validateString(responseData, "access_token", PARSE_ERROR_PREFIX);
+    int expiresInSeconds =
+        OAuth2Utils.validateInt32(responseData, "expires_in", PARSE_ERROR_PREFIX);
+    long expiresAtMilliseconds = clock.currentTimeMillis() + expiresInSeconds * 1000L;
+    return new AccessToken(accessToken, new Date(expiresAtMilliseconds));
+  }
+
+  /** Create a self-signed JWT. */
+  String createAssertion(JsonFactory jsonFactory, long currentTime, URI apiAudience)
+      throws IOException {
+
+    // Set JWT header.
+    JsonWebSignature.Header header = new JsonWebSignature.Header();
+    header.setAlgorithm("RS256");
+    header.setType("JWT");
+    header.setKeyId(privateKeyId);
+
+    // Set JWT payload.
+    JsonWebToken.Payload payload = new JsonWebToken.Payload();
+    payload.setIssuer(getIssSubValue());
+    payload.setSubject(getIssSubValue());
+    payload.setIssuedAtTimeSeconds(currentTime / 1000);
+    payload.setExpirationTimeSeconds(currentTime / 1000 + this.lifetime);
+    payload.setAudience(getTokenServerUri().toString());
+
+    // Sign the JWT by calling the existing JsonWebSignature library.
+    String assertion;
+    try {
+      payload.set("api_audience", apiAudience.toString());
+      assertion = JsonWebSignature.signUsingRsaSha256(privateKey, jsonFactory, header, payload);
+    } catch (GeneralSecurityException e) {
+      throw new IOException(
+          "Error signing service account access token request with private key.", e);
+    }
+
+    return assertion;
+  }
+
+  /** Get the issuer and subject value in the format GDCH token server required. */
+  @VisibleForTesting
+  String getIssSubValue() {
```

Comment:
this could be a private static, and receive projectId and ServiceIdentity as params

Reply:
done